doc/man/taskd.1.in
.TH taskd 1 2015-05-10 "${PACKAGE_STRING}" "User Manuals"

.SH NAME
taskd \- Taskserver

.SH SYNOPSIS
.B taskd <command> [options] [<args>]
.br
.B taskd -v|--version
.br
.B taskd -h|--help
.br
.B taskd diagnostics
.br
.B taskd validate <JSON | file>
.br
.B taskd help [<command>]

.SH DESCRIPTION
Taskserver is a sync server for Taskwarrior.

.SH COMMANDS
Taskserver supports several commands, which provide administrative capabilities.
The commands are:

.TP
.B taskd init [--data <root>] [options]
Performs a one-time initialization of the Taskserver instance.  Creates the
bare data structure.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd help <command> [options]
Provides option and syntax help for each command.

.TP
.B taskd config [options] [--data <root>] [--force] [<name> [<value>]]
Allows modification of the taskd configuration.  Can add, remove, modify or list
all configuration settings.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd server [--daemon] [options]
Starts the server in daemon or TTY mode.  While there is no interactivity, the
difference is whether taskd is attached to a TTY or not.

Note that sending the HUP signal to the taskd server causes a configuration
file reload before the next request is handled.

.TP
.B taskd add [--data <root>] org <org>
.TP
.B taskd add [--data <root>] user <org> <user-name>
Adds organizations and users.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd remove [--data <root>] org <org>
.TP
.B taskd remove [--data <root>] user <org> <uuid>
Removes organizations and users.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd suspend [--data <root>] org <org>
.TP
.B taskd suspend [--data <root>] user <org> <uuid>
Suspends organizations and users.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd resume [--data <root>] org <org>
.TP
.B taskd resume [--data <root>] user <org> <uuid>
Resumes organizations and users.
Either '\-\-data <root>' must be specified, or TASKDDATA must be set.

.TP
.B taskd diagnostics
Displays diagnostic information important when reporting bugs.

.TP
.B taskd validate <JSON | file>
Parses the quoted JSON string or JSON file.  Useful for debugging JSON that is
generated by Taskserver clients.

.TP
.B taskd api [options] <host:port> <file> [<file> ...]
Sends the Taskserver request in <file> to <host:port>. Options specified must
include ca.cert (for self-signed certs), api.cert, api.key.

.SH COMMON OPTIONS
Most commands support the following options:

.TP
.B --NAME=VALUE
Temporary configuration override.

.TP
.B --force
Does not ask for confirmation of changes.

.TP
.B --quiet
Suppresses verbose output.

.TP
.B --debug
Shows debugging information.

.TP
.B --data <root>
Many commands require the location of the data root, so either it must be
provided as shown, or via the TASKDDATA environment variable.

.SH CONFIGURATION
The Taskserver configuration file can be found in the top level of the data
root.  It is recommended that the 'taskd config' command is used to manipulate
this file.

Note that any configuration variable may be overridden on the command line using
this syntax:

  $ taskd server \-\-NAME=VALUE

In all commands but one, this will _temporarily_ override the value, but with
the 'init' command, which creates a config file, overrides are made permanent.

.SH ENVIRONMENT VARIABLES

Environment variables provide values that would otherwise be required at the
command line.

.TP
.B TASKDDATA=<root>
Most command require either '\-\-data <root>' or this environment variable.  If
both are provided, the command line overrides the environment.

.SH "CREDITS & COPYRIGHTS"
Copyright (C) 2010 \- 2018, Göteborg Bit Factory.

Taskserver is distributed under the MIT license. See
http://www.opensource.org/licenses/mit-license.php for more information.

.SH SEE ALSO
.BR taskdrc(5), taskdctl(1)

For more information regarding Taskserver, see the following:

.TP
The official site at
<http://taskwarrior.org>

.TP
The official code repository at
<https://github.com/GothenburgBitFactory/taskserver>

.TP
You can contact the project by emailing
<support@gothenburgbitfactory.org>

.SH REPORTING BUGS
.TP
Bugs in Taskserver may be reported to the issue-tracker at
<http://taskwarrior.org>



NEWS

New Features in Taskserver 1.2.0

  - When the server specified is ::, can use dual-stack IPv4/IPv6.
  - New default 2048 bits for Diffie-Hellman key exchange.
  - Server-side cipher selection is now preferred, preventing the client from
    switching to less-secure communication.
  - Removed broken SSLv3 and TLSv1.0 ciphers (POODLE) from default.
  - Removed broken 3DES cipher in CBC mode from default.
  - Removed broken RC4 cipher from default.
  - Removed broken MD5 cipher from default.
  - Integrates libshared.
  - Made the documentation consistently distinguish between user name and UUID.
  - Improved handling for I/O errors.
  - PKI scripts do not use 'which' for finding gnutls certool path

New configuration options in Taskserver 1.2.0

  - New 'dh_bits' setting allows an override to the 2048 bit default
    for the Diffie-Hellman key exchange.
  - Renamed 'client.cert' and 'client.key' to 'api.cert' and 'api.key', because
    the word 'client' implied it was to be used for all clients.

Removed features in 1.2.0

  - Removed the stubbed 'group' features.

---

While Taskserver has undergone testing on multiple platforms, bugs are sure to
remain.  If you encounter a bug, please enter a new issue at:

  https://github.com/GothenburgBitFactory/taskserver/issues

Or just send a message to:

  support@gothenburgbitfactory.org

Thank you.



src/taskd.cpp
////////////////////////////////////////////////////////////////////////////////
//
// Copyright 2010 - 2017, Göteborg Bit Factory.
//
// Permission is hereby granted, free of charge, to any person obtaining a copy
// of this software and associated documentation files (the "Software"), to deal
// in the Software without restriction, including without limitation the rights
// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
// copies of the Software, and to permit persons to whom the Software is
// furnished to do so, subject to the following conditions:
//
// The above copyright notice and this permission notice shall be included
// in all copies or substantial portions of the Software.
//
// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
// SOFTWARE.
//
// http://www.opensource.org/licenses/mit-license.php
//
////////////////////////////////////////////////////////////////////////////////

#include <cmake.h>
#include <iostream>
#include <new>
#include <cstring>
#include <stdio.h>
#include <stdlib.h>
#ifdef CYGWIN
#include <time.h>
#else
#include <sys/time.h>
#endif
#include <Datetime.h>
#include <FS.h>
#include <Color.h>
#include <shared.h>
#include <format.h>
#include <Database.h>
#include <taskd.h>
#ifdef HAVE_COMMIT
#include <commit.h>
#endif

#ifdef HAVE_LIBGNUTLS
#include <gnutls/gnutls.h>
#endif

////////////////////////////////////////////////////////////////////////////////
int main (int argc, const char** argv)
{
  int status = 0;

  // Create a vector of args.
  std::vector <std::string> args;
  for (int i = 1; i < argc; ++i)
    args.push_back (argv[i]);

  Config config;
  config.set ("verbose", "1");

  // Some options are hard-coded.
  if (args.size ())
  {
    if (args[0] == "-h" || closeEnough ("--help", args[0], 3))
      command_help (args);

    else if (args[0] == "-v" || closeEnough ("--version", args[0], 3))
    {
      Color bold ("bold");
      std::cout << '\n'
                << bold.colorize (PACKAGE_STRING)
#ifdef HAVE_COMMIT
                << " "
                << COMMIT
#endif
                << " built for "
                << osName ()
                << '\n'
                << "Copyright (C) 2010 - "
                << Datetime ().year ()
                << " Göteborg Bit Factory."
                << '\n'
                << '\n'
                << "Taskd may be copied only under the terms of the MIT license, "
                << "which may be found in the taskd source kit."
                << '\n'
                << "Documentation for taskd can be found using 'man taskd' or at "
                << "http://taskwarrior.org"
                << '\n'
                << '\n';
    }
    else
    {
      try
      {
        // Some defaults come from the environment.
        char* root_env = getenv ("TASKDDATA");
        if (root_env)
          config.set ("root", root_env);

        // Process all the options.
        std::vector <std::string> positionals;
        std::vector <std::string>::iterator arg;
        for (arg = args.begin (); arg != args.end (); ++arg)
        {
          if (closeEnough ("--data", *arg, 3))
          {
            ++arg;
            if (arg == args.end () || (*arg)[0] == '-')
              throw std::string ("ERROR: The '--data' option requireѕ an argument.");

            config.set ("root", *arg);
          }
          else if (closeEnough ("--quiet",  *arg, 3))  config.set ("verbose",      0);
          else if (closeEnough ("--debug",  *arg, 3))  config.set ("debug",        1);
          else if (closeEnough ("--force",  *arg, 3))  config.set ("confirmation", 0);
          else if (closeEnough ("--daemon", *arg, 3))  config.set ("daemon",       1);
          else if (taskd_applyOverride (config, *arg)) ;
          else                                         positionals.push_back (*arg);
        }

        // A database object interfaces to the data.
        Database db (&config);

        // The highest-level commands are hard-coded:
             if (closeEnough ("init",        args[0], 3)) command_init     (db             );
        else if (closeEnough ("config",      args[0], 3)) command_config   (db, positionals);
        else if (closeEnough ("help",        args[0], 3)) command_help     (    positionals);
        else if (closeEnough ("diagnostics", args[0], 3)) command_diag     (db             );
        else if (closeEnough ("server",      args[0], 3)) command_server   (db             );
        else if (closeEnough ("add",         args[0], 3)) command_add      (db, positionals);
        else if (closeEnough ("remove",      args[0], 3)) command_remove   (db, positionals);
        else if (closeEnough ("suspend",     args[0], 3)) command_suspend  (db, positionals);
        else if (closeEnough ("resume",      args[0], 3)) command_resume   (db, positionals);
        else if (closeEnough ("api",         args[0], 3)) command_api      (db, positionals);
        else if (closeEnough ("validate",    args[0], 3)) command_validate (    positionals);
        else
          throw format ("ERROR: Did not recognize command '{1}'.", args[0]);
      }

      catch (std::string& error)
      {
        if (error == "usage")
        {
          std::vector <std::string> no_args;
          command_help (no_args);
        }
        else
          std::cerr << error << '\n';
        status = -1;
      }

      catch (std::bad_alloc& error)
      {
        std::cerr << "Error: Memory allocation failed: " << error.what () << '\n';
        status = -3;
      }

      catch (...)
      {
        std::cerr << "Unknown error\n";
        status = -2;
      }
    }
  }
  else
    command_help (args);

  return status;
}

////////////////////////////////////////////////////////////////////////////////
